### PR TITLE
TestEngine changes

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo-devpack-dotnet/issues/945
- Added invalid types.
- Fix CallingScript issues.
- Added some storage methods.

### CODE REVIEW

**Comment by Jim8y** [comment]
Man I need a while to check this, please wait.

**Comment by Jim8y** [comment]
still suggest a moc server, instead of using online one:

Test method Neo.SmartContract.Testing.UnitTests.Storage.RpcStoreTests.TestRpcStore threw exception: 
System.Net.Http.HttpRequestException: Response status code does not indicate success: 502 (Bad Gateway).
    at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Neo.SmartContract.Testing.Storage.Rpc.RpcStore.TryGet(Byte[] key) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Storage\Rpc\RpcStore.cs:line 147
   at Neo.SmartContract.Testing.Storage.Rpc.RpcSnapshot.TryGet(Byte[] key) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Storage\Rpc\RpcSnapshot.cs:line 55
   at Neo.Persistence.SnapshotCache.GetInternal(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\SnapshotCache.cs:line 66
   at Neo.Persistence.DataCache.get_Item(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\DataCache.cs:line 68
   at Neo.Persistence.ClonedCache.GetInternal(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\ClonedCache.cs:line 43
   at Neo.Persistence.DataCache.get_Item(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\DataCache.cs:line 68
   at Neo.SmartContract.Native.PolicyContract.GetExecFeeFactor(DataCache snapshot) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\SmartContract\Native\PolicyContract.cs:line 99
   at Neo.SmartContract.ApplicationEngine..ctor(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock, ProtocolSettings settings, Int64 gas, IDiagnostic diagnostic, JumpTable jumpTable) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\SmartContract\ApplicationEngine.cs:line 172
   at Neo.SmartContract.Testing.TestingApplicationEngine..ctor(TestEngine engine, TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\TestingApplicationEngine.cs:line 65
   at Neo.SmartContract.Testing.TestEngine.Execute(Script script) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\TestEngine.cs:line 532
   at Neo.SmartContract.Testing.SmartContract.Invoke(String methodName, Object[] args) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\SmartContract.cs:line 89
   at Neo.SmartContract.Testing.Extensions.MockExtensions.<>c__DisplayClass5_0`1.<MockFunction>b__1(IInvocation invocation) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Extensions\MockExtensions.cs:line 76
   at InvokeStub_Func`2.Invoke(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
--- End of stack trace from previous location ---
   at Moq.Extensions.InvokePreserveStack(Delegate del, IReadOnlyList`1 args) in /_/src/Moq/Extensions.cs:line 184
   at Moq.MethodCall.<>c__DisplayClass23_0.<SetReturnComputedValueBehavior>b__0(IInvocation invocation) in /_/src/Moq/MethodCall.cs:line 376
   at Moq.Behaviors.ReturnComputedValue.Execute(Invocation invocation) in /_/src/Moq/Behaviors/ReturnComputedValue.cs:line 64
   at Moq.MethodCall.ExecuteCore(Invocation invocation) in /_/src/Moq/MethodCall.cs:line 230
   at Moq.Setup.Execute(Invocation invocation) in /_/src/Moq/Setup.cs:line 144
   at Moq.FindAndExecuteMatchingSetup.Handle(Invocation invocation, Mock mock) in /_/src/Moq/Interception/InterceptionAspects.cs:line 263
   at Moq.Mock.Moq.IInterceptor.Intercept(Invocation invocation) in /_/src/Moq/Interception/Mock.cs:line 17
   at Moq.CastleProxyFactory.Interceptor.Intercept(IInvocation underlying) in /_/src/Moq/Interception/CastleProxyFactory.cs:line 218
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at Castle.Proxies.NeoTokenProxy.get_TotalSupply()
   at Neo.SmartContract.Testing.UnitTests.Storage.RpcStoreTests.TestRpcStore() in C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Testing.UnitTests\Storage\Rpc\RpcStoreTests.cs:line 29
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

**Comment by shargon** [comment]
> still suggest a moc server, instead of using online one:
> 
> Test method Neo.SmartContract.Testing.UnitTests.Storage.RpcStoreTests.TestRpcStore threw exception: System.Net.Http.HttpRequestException: Response status code does not indicate success: 502 (Bad Gateway). at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode() at Neo.SmartContract.Testing.Storage.Rpc.RpcStore.TryGet(Byte[] key) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Storage\Rpc\RpcStore.cs:line 147 at Neo.SmartContract.Testing.Storage.Rpc.RpcSnapshot.TryGet(Byte[] key) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Storage\Rpc\RpcSnapshot.cs:line 55 at Neo.Persistence.SnapshotCache.GetInternal(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\SnapshotCache.cs:line 66 at Neo.Persistence.DataCache.get_Item(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\DataCache.cs:line 68 at Neo.Persistence.ClonedCache.GetInternal(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\ClonedCache.cs:line 43 at Neo.Persistence.DataCache.get_Item(StorageKey key) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\Persistence\DataCache.cs:line 68 at Neo.SmartContract.Native.PolicyContract.GetExecFeeFactor(DataCache snapshot) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\SmartContract\Native\PolicyContract.cs:line 99 at Neo.SmartContract.ApplicationEngine..ctor(TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock, ProtocolSettings settings, Int64 gas, IDiagnostic diagnostic, JumpTable jumpTable) in C:\Users\liaoj\git\neo-devpack-dotnet\neo\src\Neo\SmartContract\ApplicationEngine.cs:line 172 at Neo.SmartContract.Testing.TestingApplicationEngine..ctor(TestEngine engine, TriggerType trigger, IVerifiable container, DataCache snapshot, Block persistingBlock) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\TestingApplicationEngine.cs:line 65 at Neo.SmartContract.Testing.TestEngine.Execute(Script script) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\TestEngine.cs:line 532 at Neo.SmartContract.Testing.SmartContract.Invoke(String methodName, Object[] args) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\SmartContract.cs:line 89 at Neo.SmartContract.Testing.Extensions.MockExtensions.<>c__DisplayClass5_0`1.<MockFunction>b__1(IInvocation invocation) in C:\Users\liaoj\git\neo-devpack-dotnet\src\Neo.SmartContract.Testing\Extensions\MockExtensions.cs:line 76 at InvokeStub_Func`2.Invoke(Object, Object, IntPtr*) at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) --- End of stack trace from previous location --- at Moq.Extensions.InvokePreserveStack(Delegate del, IReadOnlyList`1 args) in /_/src/Moq/Extensions.cs:line 184 at Moq.MethodCall.<>c__DisplayClass23_0.b__0(IInvocation invocation) in /_/src/Moq/MethodCall.cs:line 376 at Moq.Behaviors.ReturnComputedValue.Execute(Invocation invocation) in /_/src/Moq/Behaviors/ReturnComputedValue.cs:line 64 at Moq.MethodCall.ExecuteCore(Invocation invocation) in /_/src/Moq/MethodCall.cs:line 230 at Moq.Setup.Execute(Invocation invocation) in /_/src/Moq/Setup.cs:line 144 at Moq.FindAndExecuteMatchingSetup.Handle(Invocation invocation, Mock mock) in /_/src/Moq/Interception/InterceptionAspects.cs:line 263 at Moq.Mock.Moq.IInterceptor.Intercept(Invocation invocation) in /_/src/Moq/Interception/Mock.cs:line 17 at Moq.CastleProxyFactory.Interceptor.Intercept(IInvocation underlying) in /_/src/Moq/Interception/CastleProxyFactory.cs:line 218 at Castle.DynamicProxy.AbstractInvocation.Proceed() at Castle.Proxies.NeoTokenProxy.get_TotalSupply() at Neo.SmartContract.Testing.UnitTests.Storage.RpcStoreTests.TestRpcStore() in C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Testing.UnitTests\Storage\Rpc\RpcStoreTests.cs:line 29 at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor) at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

I will do it later 🤣

File: src/Neo.SmartContract.Testing/TestingApplicationEngine.cs [hunk 1]
```diff
@@ -25,6 +25,42 @@ internal class TestingApplicationEngine : ApplicationEngine
...
+                var hash = Engine.OnGetCallingScriptHash?.Invoke(CurrentScriptHash);
+
+                if (hash is not null)
+                {
+                    return hash;
+                }
+
+                return base.CallingScriptHash;
+            }
+        }
```
Comment (Jim8y) [inline src/Neo.SmartContract.Testing/TestingApplicationEngine.cs:44]:
```suggestion
        public override UInt160 CallingScriptHash => Engine.OnGetCallingScriptHash?.Invoke(CurrentScriptHash) ?? base.CallingScriptHash;
```

File: src/Neo.SmartContract.Testing/TestingApplicationEngine.cs [hunk 2]
```diff
@@ -25,6 +25,42 @@ internal class TestingApplicationEngine : ApplicationEngine
...
+                var hash = Engine.OnGetEntryScriptHash?.Invoke(CurrentScriptHash);
+
+                if (hash is not null)
+                {
+                    return hash;
+                }
+
+                return base.EntryScriptHash;
+            }
+        }
```
Comment (Jim8y) [inline src/Neo.SmartContract.Testing/TestingApplicationEngine.cs:62]:
```suggestion
        public override UInt160 EntryScriptHash => Engine.OnGetEntryScriptHash?.Invoke(CurrentScriptHash) ?? base.EntryScriptHash;
```

**Review by Jim8y — DISMISSED** [review]

**Comment by Jim8y** [comment]
i have to stand out and speak for myself, check your code, you write more blank line than i did.

**Comment by Jim8y** [comment]
And I would expect more comments for methods. How about we add comments for every methods?

**Comment by shargon** [comment]
Wait for https://github.com/neo-project/neo/pull/3164 and sync submodule

**Review by vncoelho — COMMENTED** [review]
I plan to review next Monday or Tuesday

**Comment by shargon** [comment]
@Jim8y easier now

**Review by Jim8y — APPROVED** [review]

**Pull request merged**